### DISCUSSION TITLE
Constant crashing

### DISCUSSION THREAD

**Discussion by matthewfurr** [johnlindquist/kit#774] [Errors]
Is there somewhere else I can look for logs or the reason the Kit.app keeps crashing? It has happening 

The kit.log isn't showing me much.

```
...
[2022-05-28 23:58:17.681] [info] 🎛 Panel
[2022-05-28 23:58:24.623] [info] EXIT { pid: 31271, code: 0 }
[2022-05-28 23:58:24.624] [info] 31271: 🟡 exit 0. Prompt process: /Users/mfur/.kit/main/index.js
[2022-05-28 23:58:24.629] [info] 31271: 🛑 removed
[2022-05-28 23:58:24.644] [info] 🙈 Hide
[2022-05-28 23:58:24.652] [info] 🪟 Window     <--- last thing logged before crashing
[2022-05-29 00:08:01.826] [info] Protocols Prepared
[2022-05-29 00:08:01.829] [info] ☑ Enable tray
[2022-05-29 00:08:01.881] [info] right click
[2022-05-29 00:08:01.930] [info] Tray created
[2022-05-29 00:08:01.933] [info] 👁 Watch child: 31491
[2022-05-29 00:08:01.933] [info] Shortcuts Assigned
[2022-05-29 00:08:01.942] [info] Tick started
...
```

It has crashed 17 times between 6:41pm and midnight. I don't have any scheduled scripts running. Not sure what else to check.

**Comment by johnlindquist** [comment]
Hmm, try quitting Kit.app and delete the "~/.kit" dir.

Opening Kit.app again should re-trigger the sdk setup.

Another thing to try is:

Quit Kit.app

`cd ~/.kit`
`PATH=~/.kit/node/bin:$PATH npm i`

That should re-install the kit sdk npm dependencies and show if something failed to install.

Let me know if that fixes it.

## Question

Are you noticing the crash after any specific script/action?

**Reply by johnlindquist** [reply]
Also, what version of OSX are you on? And intel or arm64?

**Reply by matthewfurr** [reply]
Intel Macbook running macOS Big Sur (11.6.6)

I started with the npm install and everything looks good.

```console
~/.kit is 📦 v0.14.3 via ⬢ v18.2.0 
➜ ls

CODE_OF_CONDUCT.md	bin			data			hot			kar			main			platform		share			types
LICENSE			build			db			icons			kit.sock		node			run			sk
README.md		ci			emoji			images			lib			node_modules		script			target
api			cli			handler			index.d.ts		logos			package-lock.json	scripts			templates
assets			core			help			index.js		logs			package.json		setup			tmp


~/.kit is 📦 v0.14.3 via ⬢ v18.2.0 
➜ PATH=~/.kit/node/bin:$PATH npm i

added 443 packages, changed 1 package, and audited 1133 packages in 12s

102 packages are looking for funding
  run `npm fund` for details

12 vulnerabilities (10 high, 2 critical)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.


~/.kit is 📦 v0.14.3 via ⬢ v18.2.0 took 12s 
➜ 
```

I'll keep an eye on it and report back. Regarding your question about if it happens after a specific script or action—I have not noticed it to be anything specific. To be honest, I don't notice until I use the shortcut to open it up and nothing happens.

**Reply by matthewfurr** [reply]
Also, I did find a handful of these messages in the index.log located in my `~/.kenv/logs/index.log`

```
[2022-05-28 21:47:21.049] [info] 24307: Abandon caused exit. Provide a "onAbandon" handler to override.
...
[2022-05-28 23:07:12.602] [info] 27929: Blur caused exit. Provide a "onBlur" handler to override.
```

But they don't really line up with the crash in the kit.log sample I have in my first post. Neither of these show up between 
23:58 and 00:08.

**Reply by johnlindquist** [reply]
Yeah, the "onBlur" and "onAbandon" exits are normal.
- onBlur: clicking outside of Kit.app to end a script
- onAbandon: starting another script that interrupts and ends the previous script

I'm investigating some of the window blur/hide logic

**Reply by matthewfurr** [reply]
It did crash again after running `PATH=~/.kit/node/bin:$PATH npm i` inside the `~/.kit/` directory. So I deleted that directory and opened the app again. I'll give it some time, but so far so good. 🤞

**Reply by johnlindquist** [reply]
I just pushed out a new build (v1.17.17) which should hopefully take care of the root of the issue. If you right-click on the icon, you can check for an update.

**Reply by matthewfurr** [reply]
I updated to v1.17.17 and it crashes after I run anything: App Launcher, File Browser, my custom scripts, the Sample Snippet from examples. So I installed it on another Macbook running the same macOS Big Sur (11.6.6)  and I'm not having the same problem there. Now I just need to figure out what it is about this system that doesn't want it to work.

Here is a video opening  Kit, using the App Launcher, and it crashing after the script runs.

https://user-images.githubusercontent.com/5639047/170883144-b6474a98-1e68-437b-8d36-34a56de9ceb5.mp4

**Reply by matthewfurr** [reply]
I did find this in the Console Crash Reports: `Crashing on exception: Cannot remove an observer <_NSTouchBarFinderObservation 0x7fc70b0096f0> for the key path "nextResponder" from <ElectronNSWindow 0x7fc709e0bb90> because it is not registered as an observer.`

More info from that report:

```
Process:               Kit [8892]
Path:                  /Applications/Kit.app/Contents/MacOS/Kit
Identifier:            app.scriptkit
Version:               1.17.17 (1.17.17)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Kit [8892]
User ID:               501

Date/Time:             2022-05-29 13:34:18.771 -0400
OS Version:            macOS 11.6.6 (20G624)
Report Version:        12
Bridge OS Version:     3.0 (14Y908)
Anonymous UUID:        84E3D938-AF16-A359-2FDC-A11D5BB43AB6

Sleep/Wake UUID:       0B4AE797-B15E-4D82-98ED-D5ABA1B08A8C

Time Awake Since Boot: 3600 seconds
Time Since Wake:       2800 seconds

System Integrity Protection: enabled

Crashed Thread:        0  CrBrowserMain  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [8892]

Application Specific Information:
Crashing on exception: Cannot remove an observer <_NSTouchBarFinderObservation 0x7fc70b0096f0> for the key path "nextResponder" from <ElectronNSWindow 0x7fc709e0bb90> because it is not registered as an observer.
```

**Reply by johnlindquist** [reply]
Thanks for all the info

I'll keep hunting it down. I really hope I can reproduce it somehow. I've never seen that error before...

Does it crash if you simply open the prompt, then hit "escape" to dismiss it?

**Reply by johnlindquist** [reply]
Does your laptop that's crashing have a touchbar?

**Reply by matthewfurr** [reply]
> Does it crash if you simply open the prompt, then hit "escape" to dismiss it?

I didn't think to try that, but yes. CMD+; then Esc does cause it crash.

> Does your laptop that's crashing have a touchbar?

It does have a touchbar. It is a MacBook Pro (15-inch, 2017). It sits on a stand and the touchbar never gets used, but it does have one.

**Reply by johnlindquist** [reply]
Okay, I'm able to get it consistently crashing on two other touchbar MacBooks. Finally something concrete to look into!

**Reply by matthewfurr** [reply]
I'm honestly just glad I'm not sending you down some rabbit hole that is specific to only my setup. Let me know if there is anything I can provide or do to assist with the troubleshooting.

**Reply by johnlindquist** [reply]
Okay, I just pushed 1.17.18. It's working on my touchbar MacBooks. Plz test! 🙏

**Reply by matthewfurr** [reply]
It no longer crashes after every script or action runs, but it still crashes occasionally and the Console shows the same error as before.

```log
Process:               Kit [4568]
Path:                  /Applications/Kit.app/Contents/MacOS/Kit
Identifier:            app.scriptkit
Version:               1.17.18 (1.17.18)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Kit [4568]
User ID:               501

Date/Time:             2022-05-29 17:41:54.878 -0400
OS Version:            macOS 11.6.6 (20G624)
Report Version:        12
Bridge OS Version:     3.0 (14Y908)
Anonymous UUID:        84E3D938-AF16-A359-2FDC-A11D5BB43AB6

Sleep/Wake UUID:       1ACDBF48-4490-4441-A325-7FE91FB0CBE6

Time Awake Since Boot: 4700 seconds
Time Since Wake:       950 seconds

System Integrity Protection: enabled

Crashed Thread:        0  CrBrowserMain  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [4568]

Application Specific Information:
Crashing on exception: Cannot remove an observer <_NSTouchBarFinderObservation 0x7fe5c46a21a0> for the key path "nextResponder" from <ElectronNSWindow 0x7fe5c0d41b90> because it is not registered as an observer.
```

I tested some of the example scripts to make sure it isn't something I wrote causing it to crash. It crashed after running the dad joke script. Also, I found a new error in the kit.log. This actually took place before the dad joke crash. I can try to reproduce.

```log
[2022-05-29 17:44:44.013] [info] 🎛 Panel
[2022-05-29 17:44:47.373] [info] ↕ RESIZE: 768 x 242
[2022-05-29 17:44:49.341] [info] ↕ RESIZE: 768 x 480
[2022-05-29 17:44:51.938] [info] 🙈 Hide
[2022-05-29 17:44:51.951] [info] EXIT { pid: 5198, code: 0 }
[2022-05-29 17:44:51.951] [info] 5198: 🟡 exit 0. Prompt process: /Users/mfur/.kit/main/app-launcher.js
[2022-05-29 17:44:51.952] [info] 5198: 🛑 removed
[2022-05-29 17:44:51.956] [info] 🪟 Window
[2022-05-29 17:45:09.241] [info] /Users/mfur/.kit/cli/kenv-pull.js [object Object]
[2022-05-29 17:45:09.739] [error] Error: Error: net::ERR_TIMED_OUT
    at SimpleURLLoaderWrapper.<anonymous> (node:electron/js2c/browser_init:101:7169)
    at SimpleURLLoaderWrapper.emit (node:events:390:28)
[2022-05-29 17:45:09.739] [error] There was a problem updating the application
[2022-05-29 17:45:09.745] [error] Error: net::ERR_TIMED_OUT
    at SimpleURLLoaderWrapper.<anonymous> (node:electron/js2c/browser_init:101:7169)
    at SimpleURLLoaderWrapper.emit (node:events:390:28)
[2022-05-29 17:47:00.374] [info] /Users/mfur/.kit/setup/downloads.js [object Object]
[2022-05-29 17:47:21.169] [info] /Users/mfur/.kit/setup/downloads.js [object Object]
[2022-05-29 17:47:21.458] [info] /Users/mfur/.kit/setup/downloads.js [object Object]
[2022-05-29 17:48:20.727] [info] 🏚  main shortcut
[2022-05-29 17:48:20.727] [info] 🏃‍♀️ Run /Users/mfur/.kit/main/index.js
[2022-05-29 17:48:20.733] [info] Found: 
[2022-05-29 17:48:20.734] [info] 5199: 🏎 /Users/mfur/.kit/main/index.js 
[2022-05-29 17:48:20.736] [info] 5459: 🟢 start idle Prompt
[2022-05-29 17:48:20.758] [info] ↖ OPEN: { x: 456, y: 146, width: 768, height: 480 }
[2022-05-29 17:48:20.760] [info] 🎛 Panel
[2022-05-29 17:48:23.783] [info] Load /Users/mfur/.kenv/.widgets/corgi.html in 2
[2022-05-29 17:48:23.827] [info] 🙈 Hide
[2022-05-29 17:48:23.836] [info] 🪟 Window
```

**Reply by johnlindquist** [reply]
Dang, I have something else I can try. Testing it out now.

**Reply by johnlindquist** [reply]
Also, don't worry about it being script-specific. It's definitely the hide/active/makePanel state/logic/timing.

**Reply by johnlindquist** [reply]
I just pushed 1.17.19. It's stable on my machines. Plz test! 🙏

**Reply by matthewfurr** [reply]
1.17.19 is installed. No issues yet. I have a couple of scripts I want to work on so I will be using it regularly for the next couple of hours. I'll follow up. Thanks!

**Reply by johnlindquist** [reply]
Awesome. I hope this fixes it, otherwise I'd have to revert a feature I'd much rather keep around.

**Reply by johnlindquist** [reply]
Is it holding up OK? Need any help with any script ideas?

**Reply by matthewfurr** [reply]
I think it has been solved! I hasn't crashed like before!

I was trying not bother you with any questions, but since you asked... The script I'm working on now posts some form data and a file to a web service that responds with either a PDF, SVG, DXF, or PNG file. I'm trying to figure out the best way receive that response with the file and place it in the same directory as the input. The input came from a drop() method. This is what I have mangled together so far:

```javascript
// Name: POST TEST

import "@johnlindquist/kit";

const axios = await npm("axios");
const FormData = await npm("form-data");
const fs = await npm("fs.promises");

let wsToken = await env("WS_AUTH_TOKEN", {
  hint: md("Enter Token from Admin Tools"),
  ignoreBlur: true,
  secret: true,
});
let wsJob = await arg("Job Number");
let [input] = await drop();

const fileURL = input.path;
const fileType = input.type;
const fileName = input.name;

// Read image from disk as a Buffer
const upload = await fs.readFile(fileURL);

// Create a form and append image with additional fields
const form = new FormData();
form.append("jobNumber", wsJob);
form.append("upload", upload, fileName);

// Send form data with axios
const response = await axios
  .post("https://sub.domain.com/ws/endpoint", form, {
    headers: {
      ...form.getHeaders(),
      "Content-Type": fileType,
      "Authorization": wsToken,
    },
  })
  .then(({ data }) => {
    //this is where I am at now.
  });

//inspect(response.data)
```

I'm not a full time developer and will gladly take any criticism!

**Reply by johnlindquist** [reply]
Sounds like an awesome script!

On my phone for the night, but

`let fileDir = path.dirname(fileURL)` will give you the path to the dir of the the dropped file. So this is a common pattern:

```
let newFilePath = path.resolve(fileDir, "somefile.txt")

await writeFile(newFilePath, "foo")
```

I'll be back on my laptop in ~12 hours to help out. Also, we're just getting a discord server set up if you want to chat more real-time through this:

https://discord.gg/hCpbjBtY

**Reply by matthewfurr** [reply]
Thanks @johnlindquist. I'll see what I can come up with.

Also, thanks a bunch for looking into the crashes and resolving it so quickly. Kit.app is quickly becoming a staple for my daily work and I don't want to think about going back to work without it.

**Reply by johnlindquist** [reply]
Here's a demo of posting then writing the response:


[Open post-and-download-image in Script Kit](https://scriptkit.com/api/new?name=post-and-download-image&url=https://gist.githubusercontent.com/johnlindquist/bb785b9f0670e6c53b855598b8fbc3ce/raw/a070d9852816b56a05ef2ed2c457c7c530320d1e/post-and-download-image.ts")

```js
// Name: Post and Download Image

import "@johnlindquist/kit"

let response = await post<Buffer>(
  `https://respond-with-file-johnlindquist.vercel.app/api/hello`,
  {},
  {
    responseType: "arraybuffer",
  }
)

let imagePath = home("landscape.png")
await writeFile(imagePath, response.data)

exec(`open ${imagePath}`)
```

**Comment by johnlindquist** [comment]
Can you try running this script a couple times? It triggers a lot of the focus/blur/hide logic that's probably the cause of the crashing.

[Open crash-tester in Script Kit](https://scriptkit.com/api/new?name=quick-staking-wall&url=https://gist.githubusercontent.com/johnlindquist/e9cbd1e233ba3bf57783d708ad509d42/raw/5636555e66f3ce04a596616334086946a1ae1d51/quick-staking-wall.js")

```js
// Name: Crash Tester

import "@johnlindquist/kit"

let w = await widget(md(`{{log}}`))
await execLog(`echo 'zero'; sleep 1; echo 'one'`, (log) => {
    w.setState({log})
})

let result = await arg("Type something else")

w.setState({ log: result })

await wait(1000)

exit()
```
